### ISSUE TITLE
host given a random port

### ISSUE THREAD

**Issue by matthewp** [mholt/caddy#384]
Using this config:

```
example.net, www.example.net {
  tls off
  redir https://example.com{uri}
}
www.example.com {
  redir https://example.com{uri}
}
example.com {
  root www/example/public
}
```

The first 2 hosts, `example.net, www.example.net` should be on port 80 but are assigned 2015.

**Comment by abiosoft** [comment]
I think `tls off` is the reason for that. I know before now, Caddy defaults to port 2015 if scheme and port is not specified. With `tls off`, it disables automatic http/https and reverts to the previous behaviour.

Why not use `http://example.net, http://www.example.net` instead.

**Comment by mholt** [comment]
This should probably be patched. I'll look into it soon!

**Comment by mholt** [comment]
Okay, so... I'm not sure what to do here. The [docs](https://caddyserver.com/docs/caddyfile) (see "Addresses") suggest that 443 is the default port for hosts that don't look like localhost. But this is only because the letsencrypt package configures it this way; when you say `tls off` you disable letsencrypt from working, so it can't serve on port 443 - thus it reverts to the default port 2015.

I'm not actually sure this is a bug; probably should be documented, yes, but would you really want Caddy to deviate from its default port assignment?

**Comment by matthewp** [comment]
I disagree. `tls off` should mean "use http" and therefore use the default http port. And the default http port should be 80 like it is if you were to do http://example.com

**Comment by mholt** [comment]
Okay, but consider this: `tls off` means "don't use https" - not "use port 80".

**Comment by matthewp** [comment]
The default port of "don't use https" should be port 80. I can't think of a sane reason why it should be a random port like 2015 that no one is expecting. http is port 80 in everyone's mind.

**Comment by jungle-boogie** [comment]
I prefer caddy defaulting to a port > 1024 so it doesn't need to run with root. 

Until caddy and go support privilege separation, I recommend 2015 as the default.

**Comment by mholt** [comment]
I agree this will need more discussion before we make a major change like this. Both points are valid. One is logical and the other is intuitive (but technically complicated).

**Comment by jahmad** [comment]
First, apologize to chime in on an old issue.

While the docs suggests that 443 is the default port, it's not really accurate from user perspective. Caddy also provides redirection from http to https.

So from the outside (i know nothing about the internal) when tls is on, Caddy listen on 2 ports: 80 & 443.
With 'tls off', they would expect: -443. and yet, what they get is: -80 -443 +2015. it is confusing.

I'm agree with @matthewp here, Caddy should listen to port 80 by default, with fallback to 2015 when it's failed.

**Comment by jungle-boogie** [comment]
> While the docs suggests that 443 is the default port, it's not really accurate from user perspective. Caddy also provides redirection from http to https.

The docs I see indicate 2015 is the default when not using TLS:
https://caddyserver.com/docs/getting-started
> Open your browser to http://localhost:2015 to see it working. 

> it is confusing.

But if you run `./caddy`, the output will clearly report the site and port used:
`example.com:2015`

**Comment by jahmad** [comment]
> The docs I see indicate 2015 is the default when not using TLS:

But that's not what I'm talking about :).
I'm talking about the Addresses docs which linked by @mohlt.

> example.com              # Host: example.com, Port: 443

while maybe it's technically right, but from user's point of view it's not. it's "80 and 443".

and that, make port 2015 as the default become unintuitive.

**Comment by mholt** [comment]
Maybe the best way to explain it is this:

- The default port is 2015 in all cases (unless you change it with the `-port` flag).
- If automatic HTTPS is activated (which it is by default [if certain conditions are met](https://caddyserver.com/docs/automatic-https)), the default port for that host becomes 443. It also opens port 80 to serve HTTP redirects and ACME challenges.

I explain it that way in the docs because the site address by itself is enough to activate automatic HTTPS, effectively changing the port to something else. I agree it could be clearer; what do you think about my explanation I just gave? I'm trying to boil it down to two or three clear sentences.

**Comment by jahmad** [comment]
Yes, I understand that, and your docs is clear enough to me. but even your first response to this issue is that you'll patch it soon. I guess because it looks like the right thing to do. and I believe it's important to look right. 

Defaulting to non-standard port, and needing an explicit port for standard http, just doesn't feel right. I'm just trying to propose a compromise. I mean, how about two default ports? both 80 and 2015, depending on ability to bind low ports. 

That way, those who expect implicit port for standard http, and who expect it to just run without the privilege will get satisfied.

Maybe this is one of those cases that you gave us something nice and we ended up wanting more :). I'm sorry if for that.

**Comment by mholt** [comment]
> but even your first response to this issue is that you'll patch it soon.

Well, that's a little disingenuous to say -- keep reading, you'll see I misclassified the problem and changed my mind.

> Defaulting to non-standard port, and needing an explicit port for standard http, just doesn't feel right.

Keep in mind that root privileges (or special capabilities) are required to bind to low ports, which has to be done "explicitly" by the operator, so I think it's fair to default to a higher port that just works with no configuration needed.

AFAIK most people use Caddy in dev environments anyway, and don't want to deal with the hassle of configuring it to run low ports. That slows down the development process.

> I'm just trying to propose a compromise. I mean, how about two default ports? both 80 and 2015, depending on ability to bind low ports.

Frankly, because the point of Caddy is to make HTTPS the default (port 443), but that goal is a little different in dev environments where 2015 works just fine. Also, two default ports is confusing (as we've seen in this issue; I've done a poor job explaining it in the docs).

> Maybe this is one of those cases that you gave us something nice and we ended up wanting more :). I'm sorry if for that.

No need to be sorry; the feedback and discussion is good. I find value in it and take it into consideration at least, even if I don't bow to every wish I've received. (Caddy would be an utter mess already if I did do that, to be honest.)

So, to recap, I'm still convinced the current behavior of the program is correct, but perhaps I can clear up the documentation regarding it.

**Comment by jahmad** [comment]
> Well, that's a little disingenuous to say -- keep reading, you'll see I misclassified the problem and changed my mind.

I did say _first_ ... :)

> Also, two default ports is confusing (as we've seen in this issue; I've done a poor job explaining it in the docs).

and i believe there is no amount of docs could solve this kind of problem. the real cause for this issue is the reporter didn't read the docs. or he would know that he only need to remove the first block to get what he wanted. you already provided the redirection for free.

but it is understandable, his config is based on his intuition, based on his knowledge from other webservers.

> Frankly, because the point of Caddy is to make HTTPS the default (port 443), but that goal is a little different in dev environments where 2015 works just fine.

But your main features, what set caddy different from others -- https by default, http2, le, on-demand tls -- are much more usable on production. why not make it easier for it?

Combine this with le rate limit, people came from other webservers first instinct would be to disable tls for some of their sites. and this whole cycle will repeat again.

**Comment by jungle-boogie** [comment]
@jahmad,

I think it's pretty simple--you start caddy and it will tell you what port it's serving content on for the site.

So at this point you're not even suggesting doc patches because `there is no amount of docs could solve this kind of problem.`

What's your solution to get caddy running on ports lower than 1024 without root access?

**Comment by jahmad** [comment]
>I think it's pretty simple--you start caddy and it will tell you what port it's serving content on for the site.

In dev, yeah. not so much in production, when you use caddy with any init system of your favorite. start caddy, everything looks fine, then you find out actually it's not, then you need to look at the log.

>What's your solution to get caddy running on ports lower than 1024 without root access?

this is only important if 80 is the only default port. that's what the fallback for. to make current difficulties as irrelevant to this issue.

**Comment by mholt** [comment]
> this is only important if 80 is the only default port. that's what the fallback for. to make current difficulties as irrelevant to this issue.

But falling back to another port would cause ACME challenges to fail. It simply cannot happen. It's either port 80 or it's no port at all; the server should refuse to start.

I'll try to clear up the documentation on this and then I'll close the issue once I'm done with that.

**Comment by jahmad** [comment]
> But falling back to another port would cause ACME challenges to fail. It simply cannot happen. It's either port 80 or it's no port at all; the server should refuse to start.

Wait, I thought we're talking about when tls is off, why would ACME challenges needed here?

**Comment by mholt** [comment]
Oh, right. My bad.

Still, I'm just going to change the documentation then consider this issue closed.

**Comment by matthewp** [comment]
Hey @mholt, a lot of your reasoning has to do with dev, but I don't understand it. Why would I use a different instance of Caddy for every web project I'm developing? Why wouldn't I just stick a caddy file in my `~/dev` folder and serve the entire thing?  That's what I do.... and I absolutely *do* want port 80 for this, typing `localhost:somerandport/my/project` is lame compared to `localhost/my/project`.

**Comment by mholt** [comment]
Before Caddy, did you just run nginx or Apache or whatever as root, then?

**Comment by matthewp** [comment]
Yep.

**Comment by mholt** [comment]
You can do that with Caddy too, of course. Then you won't have to type :2015. (Although my Chrome history auto-completes it for me.)

**Comment by matthewp** [comment]
Oh, so you're saying that if running as root then 80 will be the default http port?

**Comment by mholt** [comment]
No, but you can use `-port 80` to make the default port, or just do `http://localhost` in your Caddyfile or something like that.

**Comment by matthewp** [comment]
Ok, then we're back to the discussion of, why are the defaults the chosen defaults? :-)  What is the workflow that we are favoring? 

Go ahead and do what you think is right and close this issue, but would be nice if we could continue the discussion either way.

**Comment by jahmad** [comment]
@mholt from what i can gather, you don't want to change caddy's behaviour because you don't want to make the code more complicated. that's your main concern. 

what i fail to understand is, from my point of view, it seems this change could simplify it, not make it more complicated.

when is caddy running on port 80?

currently: it's when user explicitly requested it, or when tls is managed by caddy. when is it managed? when user doesn't provide their own cert, when caddy can't bind low port, and bla bla, users need to read the docs for clarity.

i think it's too much rule, and can be simplified to: when caddy can bind low port.

or, did i missed something important here?

**Comment by mholt** [comment]
> from what i can gather, you don't want to change caddy's behaviour because you don't want to make the code more complicated. that's your main concern.

I do mind extra complexity, but more than just not wanting to change behavior, I'm not convinced it needs to be changed.

> when is caddy running on port 80?

Caddy is configurable; the answer to that question is not easy.

Here's the better question: What is the default port?

Answer: The default port is 2015.

Extended answer: Automatic HTTPS changes the port to 443 unless explicitly overwritten in the config. It also opens port 80. It does this *regardless* of the default port.

In other words, automatic HTTPS and the "default port" have nothing to do with each other.

No new arguments for changing the default port to 80 have come up; except that port 80 only sometimes works (if running as root, which we don't recommend anyway -- unless you use setcap but that only works in Linux and few developers will want to bother with that) and port 2015 will nearly always work.

So feel free to carry on the discussion, but I'm checking out of it so I can get things done.

**Comment by jahmad** [comment]
> I do mind extra complexity, but more than just not wanting to change behavior, I'm not convinced it needs to be changed.

Well, that's all what i'm trying to do. convince you. and no, i won't continue it either when i already failed :). thank you for the discussion.
